File: src/main/java/io/github/svegon/capi/event/render/tooltip/HopperTooltipCallback.java
package io.github.svegon.capi.event.render.tooltip;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.block.HopperBlock;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.world.BlockView;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@FunctionalInterface
public interface HopperTooltipCallback {
    Event<HopperTooltipCallback> EVENT = EventFactory.createArrayBacked(HopperTooltipCallback.class,
            (block, stack, world, tooltip, options) -> {}, listeners -> (block, stack, world, tooltip, context) -> {
        for (HopperTooltipCallback listener : listeners) {
            listener.appendHopperTooltip(block, stack, world, tooltip, context);
        }
    });

    void appendHopperTooltip(HopperBlock block, ItemStack stack, @Nullable BlockView world, List<Text> tooltip,
                             TooltipContext options);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/tooltip/HopperTooltipCallback.kt
package io.github.svegon.capi.event.render.tooltip

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.item.TooltipContext
import net.minecraft.text.Text
import net.minecraft.world.BlockView

fun interface HopperTooltipCallback {
    fun appendHopperTooltip(
        block: HopperBlock?, stack: ItemStack?, world: BlockView?, tooltip: List<Text?>?,
        options: TooltipContext?
    )

    companion object {
        @JvmField
        val EVENT: Event<HopperTooltipCallback> = EventFactory.createArrayBacked<HopperTooltipCallback>(
            HopperTooltipCallback::class.java,
            HopperTooltipCallback { block: HopperBlock?, stack: ItemStack?, world: BlockView?, tooltip: List<Text?>?, options: TooltipContext? -> }) { listeners: Array<HopperTooltipCallback> ->
            HopperTooltipCallback { block: HopperBlock?, stack: ItemStack?, world: BlockView?, tooltip: List<Text?>?, context: TooltipContext? ->
                for (listener in listeners) {
                    listener.appendHopperTooltip(block, stack, world, tooltip, context)
                }
            }
        }
    }
}
